presto-main/src/main/java/com/facebook/presto/metadata/TableHandle.java

```diff
@@ -48,29 +58,47 @@ public ConnectorTableHandle getConnectorHandle()
         return connectorHandle;
     }
 
-    @Override
-    public int hashCode()
+    @JsonProperty
+    public ConnectorTransactionHandle getTransaction()
     {
-        return Objects.hash(connectorId, connectorHandle);
+        return transaction;
+    }
+
+    @JsonProperty
+    public Optional<ConnectorTableLayoutHandle> getLayout()
+    {
+        return layout;
     }
 
     @Override
-    public boolean equals(Object obj)
+    public boolean equals(Object o)
```

Comment:
Note `ConnectorTableLayoutHandle` doesn't overrides `equals` and `hashCode`. This means if a `TableHandle` contains `ConnectorTableLayoutHandle`, the equality semantic over `ConnectorTableLayoutHandle` will be "referencing to the same object" -- which can be a bit surprising since it's not consistent with `TableHandle` without `ConnectorTableLayoutHandle`. 

I guess that's why in https://github.com/prestosql/presto/pull/363/commits/47ce7e62491691eb91126c68db09f5eca1e080ff, the `equals` and `hashCode` overrides are simply removed. If we never use `TableHandle` for comparison and in hash table, maybe we can also remove these two methods ? :)

Reply:
Actually, I am wondering if we should add equals in ConnectorTableLayoutHandle. We do need equality checks to determine we are getting  a rewritten plan or not, without equals, we actually have problem to determine so.  HashCode likely will not be used though.

Reply:
Per offline discussion:

- `equals` is required in `TableHandle` we need the stop criteria for Iterative Optimizer. 
- It's not clear why `equals` and `hashCode` is removed in prestosql/presto@47ce7e6, which we should dig a bit. 
- In this PR, we won't introduce `equals` and `hashCode` in `ConnectorTableLayoutHandle` and leave it as future work.